File: oop/java/src/rideshare/models/Person.java
package rideshare.models;

public class Person {
    private String name;
    private boolean onGoingRide;
    private int totalRides;

    public Person(String name) {
        this.name = name;
        this.onGoingRide = false;
        this.totalRides = 0;
    }

    public void startRide() {
        this.onGoingRide = true;
    }

    public void completeRide() {
        this.onGoingRide = false;
        this.totalRides++;
    }

    public String getName() {
        return name;
    }

    public boolean isOnGoingRide() {
        return onGoingRide;
    }

    public int getTotalRides() {
        return totalRides;
    }
}

File: oop/typescript/rideShare/Person.ts


export class Person {

  private name: string;
  private onRide: boolean;
  private totalRides: number;

  constructor(name: string) {
    this.name = name;
    this.onRide = false;
    this.totalRides = 0;
  }

  // This can loose the reference of 'this' in callback like setTimeout or asynchronous things
  markRideStarted(): void {
    this.onRide = true;
  }
  // This holds on to this but can NOT be overriden in child classes or called with super
  // startingRide = () => {}

  completeRide(): void {
    this.onRide = false;
    this.totalRides++;
  }

  getName(): string {
    return this.name;
  }

  isOnRide(): boolean {
    return this.onRide;
  }

  getTotalRides(): number {
    return this.totalRides;
  }

  toString() {
    return this.name + "(" + this.totalRides + ")"
  }
}